README.md
# Condensed Kubernetes

This is a training material to get knowledge about Kubernetes and prepare for Kubernetes certifications (CKA & CKAD). It has been initiated in 2020.

## Content

* [Basics](./docs/basics.md)
* [Installation](./docs/installation.md)
* [Projects](./docs/projects.md)
* [Releases](./docs/releases.md)
* Labs:
  * [LF - Kubernetes for Developers](./docs/labs-lfd259.md)
  * [LF - Kubernetes Fundamentals](./docs/labs-lfs258.md)
  * [Odroid setup](./docs/labs/odroid-setup.md)
  * [Raspberry PI setup](./docs/labs/raspberrypi-setup.md)
* [Examination](./docs/examination.md)
* [Recipes](./docs/recipes.md)

## Sandboxes

* [Katacoda](https://www.katacoda.com/)
  * [Kubernetes Playground](https://www.katacoda.com/courses/kubernetes/playground) (use `Shift+Inser` to paste from the clipboard in the web page interactive console)
  * [Lab Courses](https://www.katacoda.com/learn)
* [Play with Kubernetes](https://labs.play-with-k8s.com/)

## References

### Certifications

Programs:

* [Certified Kubernetes Application Developer (CKAD)](https://www.cncf.io/certification/ckad/)

<details>
  <summary>CKAD domains with their weights</summary>

  * Core Concepts 13%
  * Configuration 18%
  * Multi-Container Pods 10%
  * Observability 18%
  * Pod Design 20%
  * Services & Networking 13%
  * State Persistence 8%

</details>

* [Certified Kubernetes Administrator (CKA)](https://www.cncf.io/certification/cka/)

<details>
  <summary>CKA domains with their weights</summary>

  * Application Lifecycle Management 8%
  * Installation, Configuration & Validation 12%
  * Core Concepts 19%
  * Networking 11%
  * Scheduling 5%
  * Security 12%
  * Cluster Maintenance 11%
  * Logging / Monitoring 5%
  * Storage 7%
  * Troubleshooting 10%
</details>

General information:

* [Candidate handbook](https://docs.linuxfoundation.org/tc-docs/certification/lf-candidate-handbook)
* [Curriculum](https://github.com/cncf/curriculum)
* [Tips](https://docs.linuxfoundation.org/tc-docs/certification/tips-cka-and-ckad)

### Training courses

* Free
  * [Linux Foundation - Introduction to Kubernetes (LFS158)](https://training.linuxfoundation.org/training/introduction-to-kubernetes/)
  * [Microsoft - Kubernetes learning path](https://azure.microsoft.com/en-us/resources/kubernetes-learning-path/)
  * [Coursera - Getting Started with Google Kubernetes Engine](https://www.coursera.org/learn/google-kubernetes-engine)
  * [VMWare - KubeAcademy](https://kube.academy/)
  * [Collabnix - An Ultimate Kubernetes Hands-on Labs](https://collabnix.github.io/kubelabs/) ([collabnix/kubelabs](https://github.com/collabnix/kubelabs))
* Charged
  * [Linux Foundation - Kubernetes Fundamentals (LFS258)](https://trainingportal.linuxfoundation.org/learn/course/kubernetes-fundamentals-lfs258/course-introduction/course-information) > [Class forum](https://forum.linuxfoundation.org/categories/lfs258-class-forum)
  * [Linux Foundation - Kubernetes for Developers (LFD259)](https://trainingportal.linuxfoundation.org/learn/course/kubernetes-for-developers-lfd259/introduction/course-information) > [Class forum](https://forum.linuxfoundation.org/categories/lfd259-class-forum)

### Cloud Native Computing Foundation (CNCF)

* [cncf.io](https://www.cncf.io/) ([Youtube channel](https://www.youtube.com/channel/UCvqbFHwN-nwalWPjPUKpvTA))
  * [Annoucements](https://www.cncf.io/newsroom/announcements/)
  * [Landscape](https://landscape.cncf.io/zoom=200)
    * [Introducing The Cloud Native Landscape 2.0](https://www.cncf.io/blog/2018/03/08/introducing-the-cloud-native-landscape-2-0-interactive-edition/) - March 8, 2018

### Webinars

<details>
  <summary>CNCF webinars</summary>

  * [K8s audit logging deep dive](https://www.cncf.io/webinars/k8s-audit-logging-deep-dive/) - October 22, 2020
  * [Deploying Kubernetes to bare metal using cluster API](https://www.cncf.io/webinars/deploying-kubernetes-to-bare-metal-using-cluster-api/) - October 21, 2020
  * [The abc’s of Kubernetes security](https://www.cncf.io/webinars/the-abcs-of-kubernetes-security/) - October 21, 2020
  * [Building a Cloud-Native Technology Stack That Supports Full Cycle Development ](https://www.cncf.io/webinars/building-a-cloud-native-technology-stack-that-supports-full-cycle-development/) - September 9, 2020
  * [Arm Developer Experience Spanning Cloud, 5G and IoT](https://www.cncf.io/webinars/arm-developer-experience-spanning-cloud-5g-and-iot/) - September 8, 2020
  * [Running the next generation of cloud-native applications using Open Application Model (OAM)](https://www.cncf.io/webinars/running-the-next-generation-of-cloud-native-applications-using-open-application-model-oam/) - September 3, 2020
  * [Getting started with container runtime security using Falco](https://www.cncf.io/webinars/getting-started-with-container-runtime-security-using-falco/) - September 2, 2020
  * [Let’s untangle The Service Mesh](https://www.cncf.io/webinars/lets-untangle-the-service-mesh/) - September 1, 2020
  * [Local development in the age of Kubernetes](https://www.cncf.io/webinars/local-development-in-the-age-of-kubernetes/) - August 26, 2020
  * [Modern Software Development Pipeline: A Security Reference Architecture](https://www.cncf.io/webinars/modern-software-development-pipeline-a-security-reference-architecture/) - August 25, 2020
  * [Highly Scalable SaaS Apps on Kubernetes: Real Life Case Studies](https://www.cncf.io/webinars/highly-scalable-saas-apps-on-kubernetes-real-life-case-studies/) - September 9, 2020
  * [Comparing eBPF and Istio/Envoy for Monitoring Microservice Interactions](https://www.cncf.io/webinars/comparing-ebpf-and-istio-envoy-for-monitoring-microservice-interactions/) - August 4, 2020
  * [Event-Driven Cloud Native Workflows Use Cases and Patterns](https://www.cncf.io/webinars/event-driven-cloud-native-workflows-use-cases-and-patterns/) - July 29, 2020
  * [CNCF Member Webinar: Kubernetes Policies 101](https://www.cncf.io/webinars/kubernetes-policies-101/) - July 28, 2020
  * [The top 7 most useful Kubernetes APIs for comprehensive cloud native observability](https://www.cncf.io/webinars/the-top-7-most-useful-kubernetes-apis-for-comprehensive-cloud-native-observability/) - July 9, 2020
  * [Building Production-ready Services with Kubernetes and Serverless Architectures](https://www.cncf.io/webinars/building-production-ready-services-with-kubernetes-and-serverless-architectures/) - July 8, 2020
  * [Optimize your Kubernetes Clusters on Azure with Built-in Best Practices](https://www.cncf.io/webinars/optimize-your-kubernetes-clusters-on-azure-with-built-in-best-practices/) - July 7, 2020
  * [Best Practices for Running and Implementing Kubernetes](https://www.cncf.io/webinars/best-practices-for-running-and-implementing-kubernetes/) - June 30, 2020
  * [How Alibaba Extends K8s scheduler to support AI and big data workloads](https://www.cncf.io/webinars/how-alibaba-extends-k8s-scheduler-to-support-ai-and-big-data-workloads/) - July 15, 2020
  * [Kubernetes for storage, an overview](https://www.cncf.io/webinars/kubernetes-for-storage-an-overview/) - July 16, 2020
  * [Implementing Canary Releases on Kubernetes w/ Spinnaker, Istio, and Prometheus](https://www.cncf.io/webinars/implementing-canary-releases-on-kubernetes-w-spinnaker-istio-and-prometheus/) - July 22, 2020
  * [Kubernetes Secrets Management: Build Secure Apps Faster Without Secrets](https://www.cncf.io/webinars/kubernetes-secrets-management-build-secure-apps-faster-without-secrets/) - July 22, 2020
  * [Building application management platform with Open Application Model](https://www.cncf.io/webinars/cncf-ambassador-webinar-building-application-management-platform-with-open-application-model/) - July 22, 2020
  * [Observability of multi-party computation with OpenTelemetry](https://www.cncf.io/webinars/observability-of-multi-party-computation-with-opentelemetry/) - July 23, 2020
  
</details>

### Podcasts

* [Kubernetes Podcast from Google](https://kubernetespodcast.com/)

### Books

* [O'Reilly - Cloud Native DevOps with Kubernetes](https://www.oreilly.com/library/view/cloud-native-devops/9781492040750/)
* [O'Reilly - Kubernetes Up and Running](https://www.oreilly.com/library/view/kubernetes-up-and/9781491935668/)


docs/releases.md
# Releases

→ [Release notes](https://relnotes.k8s.io/)

Additional readings: [RisingStack - The History of Kubernetes on a Timeline](https://blog.risingstack.com/the-history-of-kubernetes/) - Jun 20, 2018

## v1.26: Electrifying

* [Announcement](https://kubernetes.io/blog/2022/12/09/kubernetes-v1-26-release/) - December 09, 2022
* [Sysdig What's new](https://sysdig.com/blog/kubernetes-1-26-whats-new) - November 30, 2022
* Main changes
  * [Introducing Validating Admission Policies](https://kubernetes.io/blog/2022/12/20/validating-admission-policies-alpha/)
  * [GA Support for Kubelet Credential Providers](https://kubernetes.io/blog/2022/12/22/kubelet-credential-providers/)

## 1.23

* [CNCF Live Webinar](https://community.cncf.io/events/details/cncf-cncf-online-programs-presents-cncf-live-webinar-kubernetes-123-release/) - January 4, 2022
* [Sysdig What's new](https://sysdig.com/blog/kubernetes-1-23-whats-new/) - November 30, 2021
* [Kubernetes Podcast from Google with Rey Lejano](https://kubernetespodcast.com/episode/167-kubernetes-1.23/) - December 8, 2021

## 1.22

* Annoucement: [Kubernetes 1.22: Reaching New Peaks](https://kubernetes.io/blog/2021/08/04/kubernetes-1-22-release-announcement/) - August 04, 2021
* CNCF Live Webinar: [Kubernetes 1.22 release](https://community.cncf.io/events/details/cncf-cncf-online-programs-presents-cncf-live-webinar-kubernetes-122-release/) - October 5, 2021
* Sysdig: [Kubernetes 1.22 – What’s new?](https://sysdig.com/blog/kubernetes-1-22-whats-new/) - July 29, 2021

## 1.21

→ [Kubernetes 1.21: Power to the Community](https://kubernetes.io/blog/2021/04/08/kubernetes-1-21-release-announcement/) - April 08, 2021

### 1.21 Content

* [Introducing Suspended Jobs](https://kubernetes.io/blog/2021/04/12/introducing-suspended-jobs/) - April 12, 2021
* [Introducing Indexed Jobs](https://kubernetes.io/blog/2021/04/19/introducing-indexed-jobs/) - April 19, 2021
* [PodSecurityPolicy Deprecation: Past, Present, and Future](https://kubernetes.io/blog/2021/04/06/podsecuritypolicy-deprecation-past-present-and-future/) - April 06, 2021

### 1.21 Articles & webinars

* [CNCF Live Webinar: Kubernetes 1.21 Release](https://community.cncf.io/events/details/cncf-cncf-online-programs-presents-cncf-live-webinar-kubernetes-121-release/) - May 20, 2021
* [sysdig - Kubernetes 1.21 – What’s new?](https://sysdig.com/blog/kubernetes-1-21-whats-new) - March 31, 2021

## 1.20

[Kubernetes 1.20: The Raddest Release](https://kubernetes.io/blog/2020/12/08/kubernetes-1-20-release-announcement/)

## 1.19

[Kubernetes 1.19: Accentuate the Paw-sitive](https://kubernetes.io/blog/2020/08/26/kubernetes-release-1.19-accentuate-the-paw-sitive/)

[Kubernetes 1.19: CNCF Webinar](https://www.cncf.io/webinars/kubernetes-1-19/)

1. Support for Kubernetes will be increased from 9 months (or 3 releases) to one year (stable)

2. There is a defined process on how API transition from Beta to GA (stable)

3. The API server will include a warning header when you use deprecated APIs (beta)

4. ".status.conditions" is standardised across objects (stable)

5. Standardise on "http://node-role.kubernetes.io" label (beta)

6. Removed the beta APIs from the conformance tests (beta)

7. Support CRI-ContainerD on Windows (beta)

8. The process to obtain the kubelet certificate and rotate it as its expiration date approaches is now stable

9. NodeRestriction admission plugin prevents Kubelets from self-setting labels within core namespaces (stable)

10. The Certificates API includes a Registration Authority to provision certificates for non-core uses (stable)

11. New kubeadm component config scheme aimed at customisable default for kubelets and core components (alpha)

12. You can include patched to your Kubeadm config (alpha)

13. A standard structure for Kubernetes log messages (alpha)

14. Redesign Event API (stable)

15. Ingress graduates to V1 (stable)

16. The new EndpointSlice API will split endpoints into several Endpoint Slice resources (beta)

17. The EndpointSlice API added a new AppProtocol field to allow application protocols to be specified for each port (beta)

18. SCTP is now supported as an additional protocol alongside TCP and UDP in Pod, Service, Endpoint, and NetworkPolicy (beta)

19. It is possible to set a pod’s hostname to its Fully Qualified Domain Name (FQDN) (alpha)

20. The Kubelet isn't expected to gather metrics anymore. This is delegated to the PodResources API (alpha)

21. The node topology manager is a kubelet component that centralizes the coordination of hardware resource assignments (beta)

22. Seccomp provides sandboxing, it reduces the actions that a process can perform which reduces a system potential attack surface (stable)

23. You can build Kubelet without Docker (stable)

24. You can define a global defaultConstraints that will be applied at cluster level to all of the pods that don’t define their own topologySpreadConstraints (alpha)

25. ComponentConfig is an ongoing effort to make component configuration more dynamic (beta)

26. With topologySpreadConstraints, you can define rules to distribute your pods evenly across your multi-zone cluster, so high availability will work correctly and the resource utilization will be efficient (beta)

27. With "PreemptionPolicy: Never" Pods can be placed in the scheduling queue ahead of lower-priority pods, but they cannot preempt other pods (beta)

28. You can run the scheduler with multiple Scheduling Profiles (beta)

29. New API for generic ephemeral inline volumes (alpha)

30. Storage Capacity Tracking prevents pods to be scheduled on nodes connected to CSI volumes without enough free space available (alpha)

31. CSI drivers can opt-in to volume ownership change (alpha)

32. Immutable Secrets and ConfigMaps (beta)

33. Azure disk in-tree driver is moved to CSI (beta)

34. vSphere disk in-tree driver is moved to CSI (beta)

[sysdig - What’s new in Kubernetes 1.19?](https://sysdig.com/blog/whats-new-kubernetes-1-19/)

## 1.18


resources/LFD259/SOLUTIONS/s_02/k8sSecond.sh
#!/bin/bash
#/* **************** LFD259:2020-06-18 s_02/k8sSecond.sh **************** */
#/*
# * The code herein is: Copyright the Linux Foundation, 2020
# *
# * This Copyright is retained for the purpose of protecting free
# * redistribution of source.
# *
# *     URL:    https://training.linuxfoundation.org
# *     email:  training@linuxfoundation.org
# *
# * This code is distributed under Version 2 of the GNU General Public
# * License, which you should have received with the source.
# *
# */
#!/bin/bash -x
## TxS 03-2020
## CKAD for 1.18.1
##
echo "  This script is written to work with Ubuntu 18.04"
echo
sleep 3
echo "  Disable swap until next reboot"
echo
sudo swapoff -a

echo "  Update the local node"
sleep 2
sudo apt-get update && sudo apt-get upgrade -y
echo
sleep 2

echo "  Install Docker"
sleep 3
sudo apt-get install -y docker.io

echo
echo "  Install kubeadm, kubelet, and kubectl"
sleep 2
sudo sh -c "echo 'deb http://apt.kubernetes.io/ kubernetes-xenial main' >> /etc/apt/sources.list.d/kubernetes.list"

sudo sh -c "curl -s https://packages.cloud.google.com/apt/doc/apt-key.gpg | apt-key add -"

sudo apt-get update

sudo apt-get install -y kubeadm=1.18.1-00 kubelet=1.18.1-00 kubectl=1.18.1-00

sudo apt-mark hold kubelet kubeadm kubectl
echo
echo "  Script finished. You now need the kubeadm join command"
echo "  from the output on the master node"
echo 



resources/LFD259/SOLUTIONS/s_05/CreateNFS.sh
#!/bin/bash
#/* **************** LFD259:2020-06-18 s_05/CreateNFS.sh **************** */
#/*
# * The code herein is: Copyright the Linux Foundation, 2020
# *
# * This Copyright is retained for the purpose of protecting free
# * redistribution of source.
# *
# *     URL:    https://training.linuxfoundation.org
# *     email:  training@linuxfoundation.org
# *
# * This code is distributed under Version 2 of the GNU General Public
# * License, which you should have received with the source.
# *
# */
sudo apt-get update && sudo apt-get install -y nfs-kernel-server

sudo mkdir /opt/sfw

sudo chmod 1777 /opt/sfw/

sudo bash -c "echo software > /opt/sfw/hello.txt"

sudo bash -c "echo '/opt/sfw/ *(rw,sync,no_root_squash,subtree_check)' >> /etc/exports"

sudo exportfs -ra

echo
echo "Should be ready. Test here and second node"
echo

sudo showmount -e localhost



resources/LFD259/SOLUTIONS/s_02/k8sMaster.sh
#!/bin/bash
#/* **************** LFD259:2020-06-18 s_02/k8sMaster.sh **************** */
#/*
# * The code herein is: Copyright the Linux Foundation, 2020
# *
# * This Copyright is retained for the purpose of protecting free
# * redistribution of source.
# *
# *     URL:    https://training.linuxfoundation.org
# *     email:  training@linuxfoundation.org
# *
# * This code is distributed under Version 2 of the GNU General Public
# * License, which you should have received with the source.
# *
# */
#!/bin/bash -x
## TxS 03-2020 
## v1.18.1 CKAD
echo "This script is written to work with Ubuntu 18.04"
sleep 3
echo
echo "Disable swap until next reboot"
echo 
sudo swapoff -a

echo "Update the local node"
sudo apt-get update && sudo apt-get upgrade -y
echo
echo "Install Docker"
sleep 3

sudo apt-get install -y docker.io
echo
echo "Install kubeadm, kubelet, and kubectl"
sleep 3

sudo sh -c "echo 'deb http://apt.kubernetes.io/ kubernetes-xenial main' >> /etc/apt/sources.list.d/kubernetes.list"

sudo sh -c "curl -s https://packages.cloud.google.com/apt/doc/apt-key.gpg | apt-key add -"

sudo apt-get update

sudo apt-get install -y kubeadm=1.18.1-00 kubelet=1.18.1-00 kubectl=1.18.1-00

sudo apt-mark hold kubelet kubeadm kubectl

echo
echo "Installed - now to get Calico Project network plugin"

## If you are going to use a different plugin you'll want
## to use a different IP address, found in that plugins 
## readme file. 

sleep 3

sudo kubeadm init --kubernetes-version 1.18.1 --pod-network-cidr 192.168.0.0/16

sleep 5

echo "Running the steps explained at the end of the init output for you"

mkdir -p $HOME/.kube

sleep 2

sudo cp -i /etc/kubernetes/admin.conf $HOME/.kube/config

sleep 2

sudo chown $(id -u):$(id -g) $HOME/.kube/config

echo "Apply Calico network plugin from ProjectCalico.org"
echo "If you see an error they may have updated the yaml file"
echo "Use a browser, navigate to the site and find the updated file"

kubectl apply -f https://docs.projectcalico.org/manifests/calico.yaml

echo
echo
sleep 3
echo "You should see this node in the output below"
echo "It can take up to a mintue for node to show Ready status"
echo
kubectl get node
echo
echo
echo "Script finished. Move to the next step"



